src/main/java/apoc/hashing/Fingerprinting.java

```diff
@@ -0,0 +1,198 @@
+package apoc.hashing;
+
+import org.neo4j.graphdb.*;
+import org.neo4j.procedure.Context;
+import org.neo4j.procedure.Description;
+import org.neo4j.procedure.Name;
+import org.neo4j.procedure.UserFunction;
+
+import javax.xml.bind.DatatypeConverter;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+public class Fingerprinting {
+
+    public static final String DIGEST_ALGORITHM = "MD5";
+
+    @Context
+    public GraphDatabaseService db;
+
+    @UserFunction
+    @Description("calculate a checksum (md5) over a node or a relationship. This deals gracefully with array properties. Two identical entities do share the same hash.")
+    public String fingerprintNodeOrRel(@Name("node_relation") Object thing, @Name(value = "propertyExcludes", defaultValue = "") List<String> excludedPropertyKeys) {
+
+        if (thing instanceof Node) {
+            return fingerprintNode((Node) thing, excludedPropertyKeys);
+        } else if (thing instanceof Relationship) {
+                return fingerprintRelationship((Relationship) thing, excludedPropertyKeys);
+        } else {
+            throw new UnsupportedOperationException("cannot fingerprint a " + thing.getClass());
+        }
+
+    }
+
+    @UserFunction
+    @Description("calculate a checksum (md5) over a the full graph. Be aware that this function does use in-memomry datastructures depending on the size of your graph.")
+    public String fingerprintGraph(@Name(value = "propertyExcludes", defaultValue = "") List<String> excludedPropertyKeys) {
+
+        return withMessageDigest(messageDigest -> {
+            // step 1: load all nodes, calc their hash and map them to id
+            Map<Long, String> idToNodeHash = db.getAllNodes().stream().collect(Collectors.toMap(
```

Comment:
can we just create a hash->node map? and use that? so we don't have to look them up again?